app/api/users.py

```diff
@@ -214,6 +214,16 @@ def before_update_object(self, user, data, view_kwargs):
         if has_access('is_admin') and data.get('deleted_at') != user.deleted_at:
             user.deleted_at = data.get('deleted_at')
 
+        if has_access('is_user_itself', user_id=user.id) and data.get('deleted_at') != user.deleted_at:
```

Comment:
@shreyanshdwivedi What's the second condition for? If a user is going to be deleted, he simply will have a `deleted_at=None` right?

Reply:
@uds5501 I've just followed the simple check which was used on deleting a user's account by admin. Just for consistency.
Though value of `deleted_at` attribute will be `None`.

Reply:
Okay!

Reply:
@shreyanshdwivedi Fix the hound violations though

Reply:
@uds5501 I don't see any violation in my updated file. I already made changes related to last 2 violations.
I also see 2 violations written beside Hound in checks but I don't see any now. Please point me if you see any violations

Reply:
`data.get('deleted_at') != user.deleted_at`

I don't understand this

Reply:
@iamareebjamal When a user's account is deleted, the `deleted_at` property is updated. So, this check ensures that the code-block below this condition is executed only when `deleted_at` is updated. Otherwise the code-block will run every-time we update any detail of the user.

Reply:
> When a user's account is deleted, the deleted_at property is updated

From where. I send a DELETE request on User endpoint, then what happens?

Reply:
Then request to update the user's details is sent. Idk exactly from where but then the code-base in `before_update_object` in `api/users.py` is called and we update the `deleted_at` attribute of user there. We've been doing same when user was deleted from `admin` too.

Reply:
`data.get('deleted_at')` This means it is being sent from user if I'm not wrong? Where does the deleted_at in data come from. Not that your implementation is wrong, but I'm trying to understand the process here

Reply:
The user has a `deleted_at` attribute. When the user is not deleted, `deleted_at` is `None` by default.
When we send a DELETE request by `destroyRecord` function on front-end, `deleted_at` data is updated to current timestamp and sent on server to update it in db. This is where this code-block comes into work. This is what I understood :sweat_smile:

Reply:
That is really really bad if it is sent from frontend. Please create an issue to mitigate it ASAP

Reply:
I'll approve and merge once this is created

Reply:
Maybe @CosmicCoder96 can help us in understanding what exactly is going on at the Front-end as I'm not really sure about this. Once confirmed by him, I'll open an issue regarding the same

Reply:
@shreyanshdwivedi Not sure if I fully get the issue, but frontend doesn't supply deleted at upon deletion server assigns that.

> When we send a DELETE request by destroyRecord function on front-end, deleted_at data is updated to current timestamp and sent on the server to update it in DB

The second part is not true, deleted_at is not sent by the frontend, rather at the server itself.

Reply:
Yes, I wanted to understand when is deleted_at assigned in data? Because `data.get('deleted_at') != user.deleted_at` seems dubious. This will be true even when undeleting the object as well. Why is there no hook of delete request and why are we doing every check in `before_update_object`, this breaks single responsibility principle, but anyway, from where does `deleted_at` come in data?

Reply:
However @iamareebjamal @shreyanshdwivedi when a user is restored, frontend does sets the value of `deleted_at` to null again.

Reply:
That is fine @CosmicCoder96

Reply:
> Why is there no hook of delete request
> from where does deleted_at come in data?

I think its handled in the library itself hence no delete hook exists.
```python
@check_method_requirements
    def delete(self, *args, **kwargs):
        """Delete an object
        """
        self.before_delete(args, kwargs)
        obj = self._data_layer.get_object(kwargs, get_trashed=(request.args.get('permanent') == 'true'))
        if obj is None:
            raise ObjectNotFound({'pointer': ''}, 'Object Not Found')
        if 'deleted_at' not in self.schema._declared_fields or request.args.get('permanent') == 'true' or current_app.config['SOFT_DELETE'] is False:
            self._data_layer.delete_object(obj, kwargs)
        else:
            data = {'deleted_at': str(datetime.now(pytz.utc))}
            self._data_layer.update_object(obj, data, kwargs)

        result = {'meta': {'message': 'Object successfully deleted'}}
        self.after_delete(result)
        return result
```
In `flask-rest-jsonapi/flask_rest_jsonapi/resource.py`.

Reply:
OK. Thanks @CosmicCoder96. Let's merge this and move discussion to a separate issue